[FLINK-8871] Support to cancel checkpoing via notification

Myasuka opened the pull request:
## What is the purpose of the change

[**update**]
Use notification to abort checkpoint and take effect in the phase of trigger/before executing sync and executing async phase of `StreamTask`.

~~Use notification to abort checkpoint, instead of previous inflexible `CancelCheckpointMarker` mechanism.~~



## Brief change log

  -  If some specific checkpoint is notified as aborted. That checkpoint-to-abort would be stored in  `StreamTask`. When `StreamTask` starts to trigger or in sync or async phase of checkpoint, the target checkpoint would be aborted.

## Verifying this change

This change added tests and can be verified as follows:

  - Add tests in `OneInputStreamTaskTest` to verify the aborted checkpoint could take effect in sync phase and async phase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **yes**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **yes**
  - If yes, how is the feature documented? **JavaDocs**

Myasuka commented:
Previous check failed due to downloading kafka tar package unsuccessfully (please refer to build [logs](https://api.travis-ci.org/v3/job/544220294/log.txt)). Rebase latest code to re-trigger again.

StephanEwen commented:
@Myasuka Thank you for working on this feature. This will be a good addition to Flink. The general mechanism makes sense to me.

I noticed a few issues in the code that would be great to address before merging.

**(1)** The trigger of checkpoints on barrier needs also to check if the checkpoint is already aborted. Similar to the triggering on RPC.
  
**(2)** I think it would be worth considering to refactor all checkpointing parts into a single place. Something like a `Checkpoints` class that captures all currently running and pending / aborted checkpoints. 

That would simplify a lot, and also make tests much easier, because tests would not need to deal with mocking the `StreamTask`.

**(3)** The PR introduces more lock synchronization, while the remaining parts are going for the mailbox instead. At least the checkpoint lock should be avoided in favor of the mailbox executor.

**(4)** We should probably rename `notifyCheckpointAbort(...)` to `notifyCheckpointAborted(...)`

**(5)** The `notifyCheckpointAbort` method in the `CheckpointListener` interface should probably not be a default method. Implementers should generally be forced to think about what to do when checkpoint is aborted.

I think given the abstractions right now, it would make sense for the `AbstractStreamOperator` to have a default implementation of that method that is empty.

Myasuka commented:
@StephanEwen thanks for your review and sorry for the reply. I have actually almost completed the refactoring checkpointing part of StreamTask locally and just crated a new issue [FLINK-14652](https://issues.apache.org/jira/browse/FLINK-14652) to track this.

BTW, what do you think of removing `lock` within notify checkpoint complete/aborted interfaces? From my side, as we still need to support legacy source, we cannot easily remove these `synchronized (lock)` in the notifications now.

StephanEwen commented:
About the lock removal: If we make actions go through the mailbox, then the mailbox implementation takes care of abstracting the backwards compatibility for legacy sources.

The fact that we have legacy sources should ideally not be visible in this code at all, otherwise we have to touch the code again when we remove the legacy sources or change the compatibility approach or locking model.

rkhachatryan commented:
Thanks for the contribution, @Myasuka .

Can you please rebase your changes to the latest master?
As we moved most checkpointing-related functionality recently from `StreamTask` to `SubtaskCheckpointCoordinator`, I think task changes should go there.

I have some high-level remarks:
1. does it make sense to split the commit (or even PR) into several pieces addressing JM and TM separately?
2. can you add a component name to the commit message?
3. `AsyncCheckpointRunnable.cancelled` seems not to be used. And why not to add a new AsyncCheckpointState instead (or reuse `DISCARDED`)?
4. I don't see how `AsyncCheckpointRunnable` is canceled or its thread is stopped; which is the original goal of this PR I believe

Myasuka commented:
@rkhachatryan Thanks for your review!
I'll rebase master code on top of https://github.com/apache/flink/pull/11721 recently.

1. > does it make sense to split the commit (or even PR) into several pieces addressing JM and TM separately?
    
I also previous single commit is a bit large, I'll split into several smaller ones.

2. > can you add a component name to the commit message?

Sure, no problem. It was my fault to missing component information.

3. > AsyncCheckpointRunnable.cancelled seems not to be used. And why not to add a new AsyncCheckpointState instead (or reuse DISCARDED)?

Previous `AsyncCheckpointRunnable.cancelled` is used for testing to verify whether this async runnable is cancelled. The reason why I not add a new `AsyncCheckpointState` is due to I reuse `AsyncCheckpointRunnable#close` to cancel the running thread which would set the state as `DISCARDED`. I think this is an open question to discuss to search for a better solution.

4. > I don't see how AsyncCheckpointRunnable is canceled or its thread is stopped; which is the original goal of this PR I believe

As I said above, I reuse `AsyncCheckpointRunnable#close` to cancel the running async checkpoint task.

Myasuka commented:
@rkhachatryan Would you please review this PR again?
I have split into two commits now and rebase with latest code.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -182,7 +181,7 @@
 
 	/** The currently active background materialization threads. */
-	private final CloseableRegistry cancelables = new CloseableRegistry();
+	private final AsyncCheckpointRunnableRegistry cancelables = new AsyncCheckpointRunnableRegistry();
```
This registry stores not only `AsyncCheckpointRunnableRegistry` but also `SubtaskCheckpointCoordinator` (and other objects).

Setting it to `AsyncCheckpointRunnableRegistry` makes me think it only stores Runnables and should go into `SubtaskCheckpointCoordinatorImpl`.

I'd suggest to explicitly keep track of async checkpointing tasks in `SubtaskCheckpointCoordinatorImpl`.

Myasuka replied:
In the beginning, I plan to wrap `CloseableRegistry` and found I have to use another lock to synchronized. That's why I currently have to extend `CloseableRegistry` to leverage the same `getSynchronizationLock()`.
And I agree the name of `AsyncCheckpointRunnableRegistry` is misleading, I'll prepare to wrap `CloseableRegistry` again as I think another internal lock should not harm much.

rkhachatryan replied:
But why not to store `AsyncCheckpointRunnable`s in `SubtaskCheckpointCoordinatorImpl`?
That is, move `Map<Long, AsyncCheckpointRunnable> checkpoints` from `AsyncCheckpointRunnableRegistry` to `SubtaskCheckpointCoordinatorImpl`.

That would make it more clear and flexible (we could change the type in map, we could avoid passing registry to `SubtaskCheckpointCoordinatorImpl` etc.).

Myasuka replied:
Because `asyncCheckpointRunnable` is registered to `closeableRegistry` when it begins to run asynchronously, we have to introduce something to wrap the `checkpointId`, `AsyncCheckpointRunnable` and `closeableRegistry` to let cancellation available, which is the role `AsyncCheckpointRunnableRegistry` played. Even we move `Map<Long, AsyncCheckpointRunnable> checkpoints` from `AsyncCheckpointRunnableRegistry` to `SubtaskCheckpointCoordinatorImpl`, we still need something to warp them all.

Moreover, since we register the `closeableRegistry` within `AsyncCheckpointRunnable`, and we create the runnable in `SubtaskCheckpointCoordinatorImpl`, I'm afraid it's not easy to avoid passing registry to `SubtaskCheckpointCoordinatorImpl`.

rkhachatryan replied:
> we still need something to warp them all.

Do you mean a scenario when all runnables need to be closed? 
For this we can explicitly iterate over a map and close them in `SubtaskCheckpointCoordinatorImpl.close()` (it's already registered with `closeableRegistry`).


> it's not easy to avoid passing registry

I didn't mean we should do it now but it's one thing we can consider in future. Having it tied to runnables makes it a bit harder.

Myasuka replied:
Actually, current `AsyncCheckpointRunnableRegistry` is a wrapper of `closeableRegistry` and ` Map<Long, AsyncCheckpointRunnable> checkpoints`.
And if we do not use a wrapper but with two separate `closeableRegistry` and `checkpoints`,  the `AsyncCheckpointRunnable` would register itself to `closeableRegistry` and put itself with checkpoint id to `checkpoints` when begin async phase. 
Once it reach the finally phase of the runnable, it will remove from the `checkpoints` and then unregister from `closeableRegistry`. Actually, I cannot see any real difference compared with current `AsyncCheckpointRunnableRegistry`.
Perhaps the name of `AsyncCheckpointRunnableRegistry` could easily mislead developers? And a name like wrapper is better?

rkhachatryan replied:
I think that `SubtaskCheckpointCoordinator` shouldn't delegate *its* responsibility of maintaining _its_ runnables to `StreamTask.closeableRegistry`. Changing the names only masquerades it.

Using the same registry creates unnecessary connection, making it more difficult to change `StreamTask` (for example, to remove registry type).

At runtime, I guess behavior is more or less the same.

Myasuka commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -1072,6 +1073,7 @@ public void run() {
 					}
 				});
 
+				sendAbortedMessages(checkpointId, pendingCheckpoint.getCheckpointTimestamp());
```
I think this is a good question. We need notify checkpoint complete to tasks ASAP if we use two-phase commit mechanism, thus `sendAcknowledgeMessages` is executed synchronously. Notify checkpoint aborted to tasks should not be so urgent and could be executed asynchronously. I'll change this behavior in this PR.

Myasuka commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointListener.java:
```diff
@@ -38,4 +38,12 @@
+	/**
+	 * This method is called as a notification once a distributed checkpoint has been aborted.
+	 *
+	 * @param checkpointId The ID of the checkpoint that has been aborted.
+	 * @throws Exception
+	 */
+	default void notifyCheckpointAborted(long checkpointId) throws Exception {}
```
I think this could still stay at `CheckpointListener.java` as this method could take effect in many places. Some state backend could leverage this method to do some cleanup things, especially those disAgg (separating storage and computation) state-backends. Moreover, we also want to introduce `notifyCheckpointSubsume` if possible in future, and too many interfaces would be a bit over design.

rkhachatryan replied:
> Some state backend could leverage this method to do some cleanup things, especially those disAgg (separating storage and computation) state-backends. 

If any of state backends will implement it, it won't be too difficult to add a call.

> too many interfaces would be a bit over design.

I think it's better to have many small interfaces rather than a bigger one (I believe it's a general recommendation).
It's also not over design here because this is an important part of the system.

As it is now, there is a lot of "dead" code with empty methods which complicates the design.

rkhachatryan replied:
As there are already such state backends, as @Myasuka mentioned below, it makes sense
(regarding multiple interfaces, they can simply be combined (with `extends`) and given to the user).

Myasuka commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java:
```diff
@@ -340,7 +340,16 @@ public void initializeState(StateInitializationContext context) throws Exception
 	@Override
 	public void notifyCheckpointComplete(long checkpointId) throws Exception {
-		stateHandler.notifyCheckpointComplete(checkpointId);
+		if (stateHandler != null) {
```
Emmm, it's just for unit test with operator not calling `initializeState`, I could try to remove this if possble.

rkhachatryan replied:
That would be nice because it makes the developer think it's also nullable in production code.

Myasuka commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandler.java:
```diff
@@ -222,6 +222,12 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
 		}
 	}
 
+	public void notifyCheckpointAborted(long checkpointId) throws Exception {
+		if (keyedStateBackend != null) {
+			keyedStateBackend.notifyCheckpointAborted(checkpointId);
+		}
+	}
```
If user develop their own state backend and rely on this method to do some necessary cleanup, this must be implemented here. Actually, that's what we rely for our internal enterprise state-backend.

rkhachatryan replied:
I highlighted the wrong lines, sorry.
I meant that there is already a default implementation in interface and there is no need in empty methods in `RocksDBKeyedStateBackend` and `HeapKeyedStateBackend`:
```
@Override
public void notifyCheckpointAborted(long checkpointId) {
	// nothing to do
}
```

Myasuka commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -93,6 +101,8 @@
 		this.unalignedCheckpointEnabled = unalignedCheckpointEnabled;
 		this.prepareInputSnapshot = prepareInputSnapshot;
 		this.closeableRegistry.registerCloseable(this);
+		this.abortedCheckpointIds = new ConcurrentSkipListSet<>();
```
Good catch, this feature is developed when `StreamTask` has no mailbox, that's why we use a `ConcurrentSkipListSet` here, will refactor to `TreeSet`.

Myasuka commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -183,6 +209,41 @@ public void notifyCheckpointComplete(long checkpointId, OperatorChain<?, ?> oper
 	}
 
+	@Override
+	public void notifyCheckpointAborted(long checkpointId, OperatorChain<?, ?> operatorChain, Supplier<Boolean> isRunning) throws Exception {
```
`notifyCheckpointAborted` is executed asynchronously, and I prefer a `Suppluer` here to provide more accurate result.

Myasuka commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -183,6 +209,41 @@ public void notifyCheckpointComplete(long checkpointId, OperatorChain<?, ?> oper
+		if (isRunning.get()) {
+			LOG.debug("Notification of aborted checkpoint for task {}", taskName);
+			// only happens when the task always received checkpoints to abort but never trigger or executing.
+			if (abortedCheckpointIds.size() >= DEFAULT_MAX_RECORDED_ABOTRED_CHECKPOINTS) {
+				abortedCheckpointIds.pollFirst();
+			}
```
This feature develops as once two very large jobs queued a lot of checkpoint requests (cannot get the checkpoint lock) on task side. The checkpoint on JM side just expire and trigger new one, while they cannot execute on task side. After nearly one day, many queue checkpoints flood as the lock is released and lead to many operators to execute checkpoint at the same time continuously, which lead to the DFS crashed and unavailable.
That's why we store those notified checkpoints, not sure whether have been explained clearly.

rkhachatryan replied:
I'm not sure I fully understood your explanation. I think you are talking about why the limit exists at all, while I meant that it should be enforced later - after the line 252:
```
abortedCheckpointIds.add(checkpointId);
```

Myasuka commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -183,6 +209,41 @@ public void notifyCheckpointComplete(long checkpointId, OperatorChain<?, ?> oper
+				abortedCheckpointIds.pollFirst();
+			}
+
+			boolean canceled = closeableRegistry.cancelAsyncCheckpointRunnable(checkpointId);
```
I think the `channelWrittenFuture` is not canceled is during normal cancel phase existed even this feature is not introduced. And storing the `Future<?>` cannot ensure canceling that future could discard those written state handles, thus we should store the `AsyncCheckpointRunnable` itself.

rkhachatryan replied:
`channelWrittenFuture` is canceled by the `ChannelStateWriter` during "normal" abort (its state handle doesn't need to be discarded). 
This PR only "speeds up" the abort, cancellation marker will follow, right?  So eventually it will be canceled too.

Any other interruptible operation (e.g. reporting to JM) should be interrupted by canceling the future returned by the `executorService`.  But this is probably out of scope of this PR.

Myasuka replied:
Thanks for your clarification, I would also try to call `ChannelStateWriter#abort` during `SubtaskCheckpointCoordinator#notifyCheckpointAborted` and add logic to verify this in unit test.

Myasuka commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorTest.java:
```diff
@@ -58,4 +83,212 @@ public void testNotifyCheckpointComplete() throws Exception {
+			operatorChain,
+			() -> true);
+		assertFalse(checkpointOperator.isCheckpointed());
+		assertEquals(0, subtaskCheckpointCoordinator.getAbortedCheckpointSize());
+		assertEquals(0, subtaskCheckpointCoordinator.getCloseableRegistry().getAsyncCheckpointRunnables().size());
```
Actually, `checkpointOperator.isCheckpointed()` has already verified that "operator snapshots are not taken".
I could add `assertEquals(-1, stateManager.getReportedCheckpointId());` to verify `TaskStateManager` has not been reported.

rkhachatryan replied:
Makes sense. 
Another thing I meant is that we shouldn't expose implementation details unnecessarily. 
So ideally we could remove `SubtaskCheckpointCoordinatorImpl.getAsyncCheckpointRunnableRegistry` and 
`AsyncCheckpointRunnableRegistry.getAsyncCheckpointRunnables`.

Myasuka replied:
Got your point, already removed this two methods and verify the logic using other ways.

Myasuka commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorTest.java:
```diff
@@ -58,4 +83,212 @@ public void testNotifyCheckpointComplete() throws Exception {
+	private <T> OneInputStreamOperator<T, T> streamOperatorWithSnapshot(OperatorSnapshotFutures operatorSnapshotResult) throws Exception {
+		@SuppressWarnings("unchecked")
+		OneInputStreamOperator<T, T> operator = mock(OneInputStreamOperator.class);
+		when(operator.getOperatorID()).thenReturn(new OperatorID());
+
+		when(operator.snapshotState(anyLong(), anyLong(), any(CheckpointOptions.class), any(CheckpointStreamFactory.class)))
+			.thenReturn(operatorSnapshotResult);
```
Thanks for you suggestion, I'll enrich `CheckpointOperator`.

Myasuka commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskLocalStateStoreImplTest.java:
```diff
@@ -153,6 +153,21 @@ public void confirmCheckpoint() throws Exception {
+		final int aborted = chkCount - 2;
+		List<TaskStateSnapshot> taskStateSnapshots = storeStates(chkCount);
+		taskLocalStateStore.abortCheckpoint(aborted);
+		checkPrunedAndDiscarded(taskStateSnapshots, aborted, aborted + 1);
+		checkStoredAsExpected(taskStateSnapshots, 0, aborted);
+		checkStoredAsExpected(taskStateSnapshots, aborted + 1, chkCount);
```
`taskLocalStateStore.retrieveLocalState(checkpointId)` would actually already check snapshot was removed from `TaskLocalStateStoreImpl.storedTaskStateByCheckpointID`.

Myasuka commented:
@flinkbot run azure

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -60,17 +63,22 @@
 	private final StreamTaskActionExecutor actionExecutor;
 	private final boolean unalignedCheckpointEnabled;
 	private final BiFunctionWithException<ChannelStateWriter, Long, CompletableFuture<Void>, IOException> prepareInputSnapshot;
+	/** The IDs of the checkpoint for which we are notified aborted. */
+	private final NavigableSet<Long> abortedCheckpointIds;
```
Why do we need to track the aborted checkpoints?

Myasuka replied:
The notification of aborted checkpoints could come before task start to execute the checkpoint.
We store aborted checkpoints so that we could avoid to execute any sync or async phase of checkpoint on task side when we know this checkpoint has been aborted.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -183,6 +235,42 @@ public void notifyCheckpointComplete(long checkpointId, OperatorChain<?, ?> oper
+				abortedCheckpointIds.pollFirst();
+			}
+
+			channelStateWriter.abort(checkpointId, new NotifiedCheckpointAbortedException(checkpointId));
```
`try/catch`?
same for the operators

Myasuka replied:
Would you please give more clear explanations for this? Did you mean we cannot call `channelStateWriter.abort` twice. Moreover, what do you mean for `same for the operators`?

rkhachatryan replied:
Sorry for not being clear. We shouldn't call `channelStateWriter.abort` before receiving a cancelation marker.

> Moreover, what do you mean for same for the operators?

Added a separate comment.

Myasuka replied:
If the async checkpoint on task side is aborted via RPC message, and we did not call `channelStateWriter.abort`. If above stream did not broadcast a cancellation marker to downstream, what happened to `channelStateWriter` as that is not called as aborted.

rkhachatryan replied:
In that case, `CheckpointBarrierHandler` will call abort on the next barrier.
Though I don't think it's possible that task skips the cancellation marker.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapKeyedStateBackend.java:
```diff
@@ -311,6 +311,11 @@ public void notifyCheckpointComplete(long checkpointId) {
 		//Nothing to do
 	}
 
+	@Override
+	public void notifyCheckpointAborted(long checkpointId) {
+		// nothing to do
+	}
+
```
(or the other way around, as @StephanEwen mentioned: no default implementation)

Myasuka replied:
I forget Stephan's suggestion to remove the default implementation, I'll update to remove the default implementation of `notifyCheckpointAborted`

rkhachatryan commented:
Thanks for the updates @Myasuka 

There are still some questions open, please see the comments above.

Besides that, can you please list some IT cases covering checkpoint cancelation in the description?
 
I'm a bit concerned about the following scenarios:
1. a barrier arrives for an already canceled checkpoint; alignment still happens as usual but the result is discarded, right?
2. recently, upstream tasks started to block in aligned mode (#11351). With this change, they still will be unblocked by the downstream despite of cancelation via RPC, right?

Myasuka commented:
@rkhachatryan thanks for your review. It's quite late for me at my time now and plan to update at my day time.
Before go to sleep, I could answer the 1st question: barrier alignment would still happen as usual, current PR would not affect barrier align part as discussed in the beginning, and we could improve this in the future.

Myasuka commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java:
```diff
@@ -464,6 +464,11 @@ public void notifyCheckpointComplete(long completedCheckpointId) throws Exceptio
 		}
 	}
 
+	@Override
+	public void notifyCheckpointAborted(long checkpointId) {
+		// nothing to do
+	}
+
```
I prefer the idea to not leave default implementation of `notifyCheckpointAborted`, thus leave it here could be acceptable.

Myasuka commented:
For the 2nd question @romanovacca , after discussed with @zhijiangW offline, we believe this PR with cancellation via RPC would not take part in the data transmission in channel, thus the upstream task would be unblocked once the channel is unblocked which is the same logic as before.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -183,6 +235,42 @@ public void notifyCheckpointComplete(long checkpointId, OperatorChain<?, ?> oper
+			}
+
+			for (StreamOperatorWrapper<?, ?> operatorWrapper : operatorChain.getAllOperators(true)) {
+				operatorWrapper.getStreamOperator().notifyCheckpointAborted(checkpointId);
```
What if `notifyCheckpointAborted` fails for one operator, should we continue to the next?

Myasuka replied:
I think this is the same for `notifyCheckpointComplete`, catch the exception and continue to the next operator should be a solution.

rkhachatryan replied:
For `notifyCheckpointComplete` it's less important because operators are expected to work correctly even without the notification (and `channelStateWriter` has only in-memory resources).

Another question: this is a new method and we call it only on RPC notification, right? Should we also call it in `abortCheckpointOnBarrier`?

Myasuka commented:
I think current IT cases do not target to verify the cancel notification, I'll create one to target to veriy this.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -183,6 +235,42 @@ public void notifyCheckpointComplete(long checkpointId, OperatorChain<?, ?> oper
+			LOG.debug("Notification of aborted checkpoint for task {}", taskName);
+			// only happens when the task always received checkpoints to abort but never trigger or executing.
+			if (abortedCheckpointIds.size() >= maxRecordAbortedCheckpoints) {
+				abortedCheckpointIds.pollFirst();
```
Why do we remove the oldest checkpoint? Why not the one that was _canceled first_?
I guess both are valid options, but for the 2nd we could use `LinkedHashMap` with O(1) time complexity.

Myasuka replied:
I think this suggestion sounds reasonable, changed to `ArrayDeque`.

rkhachatryan commented on flink-tests/src/test/java/org/apache/flink/test/checkpointing/NotifyCheckpointAbortedITCase.java:
```diff
@@ -0,0 +1,576 @@
+
+		// let the checkpoint-1 failed finally.
+		TestingCompletedCheckpointStore.abortCheckpointLatch.trigger();
+		while (deadline.hasTimeLeft()) {
```
I think this check will pass if the deadline is missed. Should it fail instead?

rkhachatryan commented on flink-tests/src/test/java/org/apache/flink/test/checkpointing/NotifyCheckpointAbortedITCase.java:
```diff
@@ -0,0 +1,576 @@
+
+		// verify BeforeExecuteCheckpointSink never execute checkpoint-1.
+		while (deadline.hasTimeLeft()) {
+			if (!BeforeExecuteCheckpointSink.snapshotIds.isEmpty()) {
```
Can we use accumulators instead of a static field?
It could save from sync/wait and improve readability (no static field).

rkhachatryan commented on flink-tests/src/test/java/org/apache/flink/test/checkpointing/NotifyCheckpointAbortedITCase.java:
```diff
@@ -0,0 +1,576 @@
+	 *                                  |--> BarrierDelayMap -> BeforeExecuteCheckpointSink
+	 */
+	@Test(timeout = TEST_TIMEOUT)
+	public void testNotifyCheckpointAborted() throws Exception {
```
This test does a great job of testing the functionality :+1: 

However, considering its complexity and unit-test coverage, I think a valid trade-off would be to simply test that operator is notified that the checkpoint is aborted.

Myasuka commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointListener.java:
```diff
@@ -38,4 +38,12 @@
+	 * @param checkpointId The ID of the checkpoint that has been aborted.
+	 * @throws Exception
+	 */
+	void notifyCheckpointAborted(long checkpointId) throws Exception;
```
Once this PR merged, I planned to add release notes in JIRA. From my previous experience, this interface change might not publish in the final release notes of flink official website.

pnowojski commented:
There is a test failure:
```
SubtaskCheckpointCoordinatorTest.testNotifyCheckpointAbortedDuringAsyncPhase:195 » IllegalArgument
```

Myasuka commented:
@pnowojski thanks for your remind.
I think I have addressed all comments to update the commits.

Since current CI is still pending, my local pipeline based on the same commit [d023c50](https://github.com/apache/flink/commit/d023c507a423b31f38203581c8eed27d28cb3a3c) get a [green result](https://myasuka.visualstudio.com/flink/_build/results?buildId=60&view=results).

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnables.java:
```diff
@@ -0,0 +1,85 @@
+ * Track the {@link AsyncCheckpointRunnable}s in {@link SubtaskCheckpointCoordinator}.
+ */
+@Internal
+public class AsyncCheckpointRunnables implements Closeable {
```
I doubt whether we need this class, given that:
1. the only user is `SubtaskCheckpointCoordinatorImpl`
2. it's lifecycle is bound to one of `SubtaskCheckpointCoordinatorImpl`
3. the only field is a map (after removing the lock)

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -126,6 +162,15 @@ public ChannelStateWriter getChannelStateWriter() {
+		return abortedCheckpointIds.size();
+	}
+
+	private boolean checkpointAlreadyAborted(long checkpointId) {
```
~~nit~~ ambiguous name (not `nit` because I indeed suggested a wrong name ))

`getAndClearAbortedStatus`?

Myasuka replied:
I think `getAndClearAbortedStatus` did not give explicit description of the result when checking the notified checkpoint-id. I think `isCheckpointAlreadyAborted` might be better than original name. 

Moreover, I think we don't need to explain that we will clear the aborted checkpoint-id if found which is too technically.

rkhachatryan replied:
By convention, `isXxx` should be used for boolean getters.
I think it's always good to give self-explaining names to methods to prevent surprises and looking into their code.

Myasuka replied:
How about `checkAndClearAbortedStatus` as I need to return boolean result.

rkhachatryan replied:
:+1:

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -142,6 +187,13 @@ public void checkpointState(
 		// checkpoint alignments
 
+		// Step (0): Record the last triggered checkpointId.
+		lastCheckpointId = metadata.getCheckpointId();
```
`lastCheckpointId = Math.max(lastCheckpointId, metadata.getCheckpointId());`
?

Myasuka replied:
`checkpointState` is executed in sync phase of checkpoint on task which should behave as a strictly increasing order due to the barrier align mechanism.

rkhachatryan replied:
It can also be triggered through RPC (for sources).
I agree that it shouldn't be possible now, but I think it's good to have such a guard or check to protect against bugs or changes.

Myasuka replied:
If so, the condition below have to change which lead to I have to record a lot of unnecessary checkpoint ids.

~~~java
if (checkpointId > lastCheckpointId) {
	// only record checkpoints that have not triggered on task side.
       abortedCheckpointIds.add(checkpointId);
}
~~~

rkhachatryan replied:
We can just force the assumption with
```Preconditions.checkArgument(lastCheckpointId < metadata.getCheckpointId());```

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -328,6 +421,7 @@ private OperatorSnapshotFutures buildOperatorSnapshotFutures(
 
 	@Override
 	public void close() throws IOException {
+		asyncCheckpointRunnables.close();
```
`try/catch` (or guava Closer or some other way to close both no matter what?)

Myasuka replied:
OK, I'll ensure both the closeables could be closed.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -384,4 +478,12 @@ private static OperatorSnapshotFutures checkpointStreamOperator(
 		}
 	}
+
+	private static class NotifiedCheckpointAbortedException extends Exception {
```
Couldn't find any usages (remove?).

Myasuka replied:
good catch, this is something implemented in previous commits but already useless.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -126,6 +162,15 @@ public ChannelStateWriter getChannelStateWriter() {
+	}
+
+	private boolean checkpointAlreadyAborted(long checkpointId) {
+		return abortedCheckpointIds.remove(checkpointId);
```
This is `O(n)`. 
I [suggested](https://github.com/apache/flink/pull/8693#discussion_r425881776) to use `LinkedHashMap`.

Myasuka replied:
Thanks for your remind.
I'll use a set based on `LinkedHashMap` to implement a set which would remove oldest checkpoint id automatically.
~~~java
private Set<Long> createAbortedCheckpointSetWithLimitSize(int maxRecordAbortedCheckpoints) {
	return Collections.newSetFromMap(new LinkedHashMap<Long, Boolean>() {
		private static final long serialVersionUID = 1L;

		@Override
		protected boolean removeEldestEntry(Map.Entry<Long, Boolean> eldest) {
			return size() > maxRecordAbortedCheckpoints;
		}
	});
}
~~~

Myasuka commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:
```diff
@@ -179,6 +179,13 @@ public void notifyCheckpointComplete(long completedCheckpointId) {
+	@Override
+	public void notifyCheckpointAborted(long abortedCheckpointId) throws Exception {
+		synchronized (materializedSstFiles) {
+			materializedSstFiles.keySet().removeIf(checkpointId -> checkpointId == abortedCheckpointId);
```
Good catch

Myasuka replied:
Moreover, I'll implement `RocksDBKeyedStateBackend#notifyCheckpointAborted`

Myasuka commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnables.java:
```diff
@@ -0,0 +1,85 @@
+	/** Lock that guards state of this registry. **/
+	private final Object lock;
+
+	public AsyncCheckpointRunnables() {
```
I think this is something related to code style and no such guide in Flink official code-style guide. I prefer explicit constructor.

Myasuka commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnables.java:
```diff
@@ -0,0 +1,85 @@
+
+	public void registerAsyncCheckpointRunnable(long checkpointId, AsyncCheckpointRunnable asyncCheckpointRunnable) {
+		synchronized (lock) {
+			checkpoints.put(checkpointId, asyncCheckpointRunnable);
```
> Can we add a check that we don't already have this checkpointId?
The checkpoint ids should be unique on task side, and I think adding a check here could be safer, if the `checkpointId` already registered, we should throw exception out.

> Ideally, we should also check that this registry was not closed yet
I think this is reasonable, which is similar to `AbstractCloseableRegistry`.

Myasuka commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnables.java:
```diff
@@ -0,0 +1,85 @@
+ */
+@Internal
+public class AsyncCheckpointRunnables implements Closeable {
+	private final Map<Long, AsyncCheckpointRunnable> checkpoints;
+	/** Lock that guards state of this registry. **/
+	private final Object lock;
```
If we add check to `closed`, we could still need the `lock` instead of introducing `ConcurrentMap`.

rkhachatryan replied:
But then this check can (should) be moved to `SubtaskCheckpointCoordinatorImpl`.

rkhachatryan replied:
Yes, you're right.
And we have to add this check to maintain current behavior.
(both these fields however can be moved to `SubtaskCheckpointCoordinatorImpl` as I suggested [here](https://github.com/apache/flink/pull/8693#discussion_r427145436)).

Myasuka replied:
I think your suggestion make sense, and the only other place using it is `LocalStateForwardingTest#testReportingFromSnapshotToTaskStateManager` ans I will pass a `SubtaskCheckpointCoordinatorImpl` to the created `AsyncCheckpointRunnable`.

carp84 commented:
Already merged into master via:
* f0ed29c06d331892a06ee9bddea4173d6300835d
* fae6a6cad3f4fe30d80cc9dd664b2efc72c0be36
* 1f6b1e6b8ffb38f196c2a1c6348cb96935e9a9cc

release-1.11 via:
* fa731288518c8ebf66f40b4e0e9b1929546b6257
* fcacc42e17f00cb47c5c16fe75af035f784ae1fa
* 6591263c8871898f602d122aac336f2ad63bbb1c
